File: 2022/python/04.py
data = open("../data/04-1.txt").read().split("\n")

def findOverlap(myPair, part):
  first, second = myPair[0].split('-'), myPair[1].split('-')
  if part == '1':
    if int(first[0]) <= int(second[0]) and int(first[1]) >= int(second[1]) or int(first[0]) >= int(second[0]) and int(first[1]) <= int(second[1]):
      return 1
  elif part == '2':
    if (int(first[0]) >= int(second[0]) and int(second[1]) >= int(first[0])) or (int(second[0]) >= int(first[0]) and int(second[0]) <= int(first[1])):
      return 1
  return 0

count1, count2 = 0, 0
for line in data:
  myPair = line.split(",")
  count1 += findOverlap(myPair, '1')
  count2 += findOverlap(myPair, '2')

# part 1 answer
print(count1)
# part 2 answer
print(count2)

File: 2022/go/04.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"regexp"
	"strconv"
)

func main() {
	file, _ := os.Open("../data/04-1.txt")
	scanner := bufio.NewScanner(file)
	regExp, _ := regexp.Compile("(\\d+)-(\\d+),(\\d+)-(\\d+)")

	count1, count2 := 0, 0
	var a, b, c, d int

	for scanner.Scan() {
		matches := regExp.FindStringSubmatch(scanner.Text())
		a, _ = strconv.Atoi(matches[1])
		b, _ = strconv.Atoi(matches[2])
		c, _ = strconv.Atoi(matches[3])
		d, _ = strconv.Atoi(matches[4])
		if a <= c && b >= d || a >= c && b <= d {
			count1++
		}
		if a >= c && d >= a || c >= a && c <= b {
			count2++
		}
	}
	// part 1 answer
	fmt.Println(count1)
	// part 2 answer
	fmt.Println(count2)
}